### ISSUE TITLE
home: account pam_systemd_home.so - remove the if (PAM_USER_UNKNOWN) return PAM_SUCCESS test

### ISSUE THREAD

**Issue by chandradeepdey** [systemd/systemd#16906]
Currently `account pam_systemd_home.so` returns `PAM_SUCCESS` on `PAM_USER_UNKNOWN`. Programs like `login` depend on the value returned by `pam_acct_mgmt()` to block a user with an expired password from login.

This is hard to get right with the current behaviour, since any error values from `account pam_unix.so` with a `ignore` action (like in the `sufficient` example from the manpage) results in `account pam_systemd_home.so` just sending a `PAM_SUCCESS`. Also, this forces the order of the modules to be in a certain way.

This is an alternative I found -
```
account    [success=ok new_authtok_reqd=ok ignore=1 user_unknown=1 default=bad] pam_unix.so
account    [default=1]             pam_permit.so
# if reached here due to user_unknown, this will return success.
# but an actual unknown user is the fault of the application.
# account expects the user to be already authenticated and it is impossible
# that an user is already authenticated but is unknown to both account
# modules.
account    required                pam_systemd_home.so
account    required                pam_time.so
```
It might still be better to just return `PAM_USER_UNKNOWN`, so the sysadmin can test against that value in the pam configuration if they want to. Doing this change will also allow the modules to be placed in any order.

I opened #15116 and associated PR #15118 quite a while ago, where I first changed to reporting `PAM_USER_UNKNOWN` as is, and then force pushed to reporting `PAM_IGNORE` instead. I closed those because, on further reading, I thought that the current design was good enough, I never considered expiring passwords and stuff, and I knew less about PAM and tried to apply my `PAM_IGNORE` solution uniformly to every function. The branch was deleted long ago, so starting fresh with this.

Also: Some [downstream discussion](https://bugs.archlinux.org/task/65819), loqs was the one who wanted the return value to be truthful and posted a patch there too.

**Comment by chandradeepdey** [comment]
The "alternative" i found doesn't even work. `account pam_unix.so` has a pretty bad `user_unknown` test which just does `pam_get_item(PAM_USER)`. Obviously for a systemd-homed user the username won't be empty, and so PAM_USER_UNKNOWN is never returned. some other error is returned by some later test.

Removing the account test here would allow configurations like
```
#%PAM-1.0

-auth      [success=3 ignore=1 user_unknown=1 default=bad]  pam_systemd_home.so
auth       [default=1]                 pam_permit.so
auth       [success=1 ignore=ignore default=bad]  pam_unix.so          try_first_pass nullok
auth       [default=die]               pam_faillock.so      authfail
auth       optional                    pam_permit.so
auth       required                    pam_faillock.so      authsucc
auth       required                    pam_env.so

-account   [success=ok new_authtok_reqd=ok ignore=1 user_unknown=1 default=bad]  pam_systemd_home.so
account    [default=1]                 pam_permit.so
account    required                    pam_unix.so
account    required                    pam_time.so

-password  [success=ok ignore=1 user_unknown=1 default=bad]  pam_systemd_home.so
password   [default=1]                 pam_permit.so
password   required                    pam_unix.so          try_first_pass nullok shadow

-session   optional                    pam_systemd_home.so
session    required                    pam_limits.so
session    required                    pam_unix.so
```

**Comment by chandradeepdey** [comment]
```
# success can be due to user_unknown or actually succeeding. don't trust it
account    [success=1 new_authtok_reqd=ok ignore=1 default=bad]  pam_systemd_home.so
account    [default=1]                 pam_permit.so
# authinfo_unavail can be either due to a homed user, or a genuine authinfo_unavail
account    [success=ok new_authtok_reqd=ok authinfo_unavail=ignore ignore=1 default=bad]  pam_unix.so
# turn authinfo_unavail=ignore to success=ok. user should be already
# authenticated, so we don't care if this success is incorrect
account    optional                    pam_permit.so
account    required                    pam_time.so
```
here's a hacky workaround if anyone is interested.

**Comment by poettering** [comment]
Hmm, so this is the "account" stack we suggest in the man page of pam_systemd_home:

```
           account   required   pam_nologin.so
           -account  sufficient pam_systemd_home.so
           account   sufficient pam_unix.so
           account   required   pam_permit.so
```

are you saying that it is broken, because the "sufficient" means that pam_systemd_home.so saying "no" will be ignored, and since it returns "sucess" for all user's it doesn't know noone else can say "no" either? that'd be quite a bad default suggested config I must admit. Did i grok this right though?

i am not entirely sure what the right set of return values is here though and what the right suggested stack would be... and even further: for the other three stacks...

If we make "account [success=done user_unknown=ignore default=bad] pam_systemd_home.so" the correct suggested line, and return all errors as they are we are good, no?

**Comment by chandradeepdey** [comment]
Yes. The one suggested in the manpage is probably worse in returning a proper error to the applications. It just guarantees that line 3 is never reached if the user is not a systemd-homed managed user.

Without removing the test there is really no correct suggested stack, since both pam_unix.so and pam_systemd_home.so return ambiguous results. At least one of them needs to return a correct result. I opened an issue in linux-pam too, but I think that will require a major rework that they probably won't do. The one in the previous comment tries to preserve most return values, but it still has the gap where I turn `authinfo_unavail` into `success`.

With the test removed https://github.com/systemd/systemd/issues/16906#issuecomment-683439776 can be a suggested stack. I will update the comment to include some comments.

`account [success=done user_unknown=ignore default=bad] pam_systemd_home.so` will work with a separate `substack` for `pam_systemd_home.so` and `pam_unix.so`. At least on Arch Linux those two are followed by more `account` entries and the file is `include`d in other files that have their own `account` entries that won't be executed on `success=done`.

**Comment by chandradeepdey** [comment]
@poettering any progress on this?

**Comment by chandradeepdey** [comment]
Fixed in #17376
